README.md
| ![IE](https://github.com/alrra/browser-logos/blob/master/src/edge/edge_48x48.png?raw=true) | ![Chrome](https://github.com/alrra/browser-logos/blob/master/src/chrome/chrome_48x48.png?raw=true) | ![Firefox](https://github.com/alrra/browser-logos/blob/master/src/firefox/firefox_48x48.png?raw=true) | ![Opera](https://github.com/alrra/browser-logos/blob/master/src/opera/opera_48x48.png?raw=true) | ![Safari](https://github.com/alrra/browser-logos/blob/master/src/safari/safari_48x48.png?raw=true) |
| ------------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------- |
| IE 10+ ✔                                                                                   | Chrome 31.0+ ✔                                                                                     | Firefox 31.0+ ✔                                                                                       | Opera 30.0+ ✔                                                                                   | Safari 7.0+ ✔                                                                                      |

![NPM version](http://img.shields.io/npm/v/xy-time-select.svg?style=flat-square)
![node version](https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square)
![npm download](https://img.shields.io/npm/dm/xy-time-select.svg?style=flat-square)

[![xy-time-select](https://nodei.co/npm/xy-time-select.png)](https://npmjs.org/package/xy-time-select)

# xy-time-select

时间选择器组件组件, 可从下拉列表选择时间，也可以直接输入

## 安装

```bash
# yarn
yarn add xy-time-select
```

## 使用例子

```tsx
import React from "react";
import ReactDOM from "react-dom";
import TimeSelect from "xy-time-select";
ReactDOM.render(<TimeSelect />, container);
```

## API

| 属性           | 说明                                        | 类型                                        | 默认值  |
| -------------- | ------------------------------------------- | ------------------------------------------- | ------- |
| disabled       | 是否禁用                                    | boolean                                     | false   |
| value          | 输入框值, 比如 01:00                        | string                                      | 无      |
| defaultValue   | 输入框默认值                                | string                                      | 无      |
| placeholder    | 占位符文本                                  | string                                      | 无      |
| autoFocus      | 自动焦点                                    | boolean                                     | 无      |
| onChange       | 输入框 change 事件                          | (value: string) => void                     | 无      |
| popupClassName | 弹出内容类名                                | string                                      | 无      |
| stretch        | 下拉列表是否宽度与输入框宽度对齐            | boolean                                     | true    |
| backfill       | 是否键盘输入完再回填                        | boolean                                     | false   |
| delay          | 防抖毫秒                                    | number                                      | 200     |
| dataSource     | 数据源                                      | OptionConfig[] / string[] / React.ReactNode | 无      |
| customItem     | 自定义渲染                                  | (props: CustomItemProps) => JSX.Element     | 无      |
| onChange       | 改变回调                                    | (value: string) => void                     | 无      |
| start          | 开始时间,时间选择器开始时间范围, 比如 09:00 | string                                      | "00:00" |
| end            | 结束时间,时间选择器开始时间范围, 比如 18:00 | string                                      | "23:59" |
| step           | 间隔时间                                    | string                                      | "00:30" |
| min            | 最小可选时间                                | string                                      | 无      |
| max            | 最大可选时间                                | string                                      | 无      |

## 开发

```sh
yarn run start
```

## 例子

http://localhost:6006

## 测试

```
yarn run test
```

## 开源许可

xy-time-select is released under the MIT license.


CHANGELOG.md
# Change Log

## 0.2.0 (Mon Dec 30 2019)

-   统一国际化实现

## 0.1.0 (Tue Nov 19 2019)

-   修复`react`与`react-dom`版本不一致导致的问题
-   测试库迁移到`@testing-library/react"`

## 0.0.42 (Wed Sep 04 2019)

-   清除控制台打印, 优化代码警告

## 0.0.41 (Thu May 30 2019)

-   剥离`date`工具函数到`utils-dom`外部库中

## 0.0.4 (Mon May 13 2019)

-   更新依赖

## 0.0.3 (Thu May 09 2019)

-   更新依赖, 更流畅的动画
-   增加 onBlur 事件
-   `icon`图标可以点击清空

## 0.0.2 (Thu May 09 2019)

-   基本实现

## 0.0.1 (Wed May 08 2019)

-   初始化项目


src/useValue.ts
import React, { useEffect, useRef, useState } from "react";
import { isShortTime } from "utils-dom";
import { DefineDefaultValue } from "utils-hooks";
import { TimeSelectProps } from "./interface";

export default function useValue(props: TimeSelectProps): [string, React.Dispatch<React.SetStateAction<string>>, (val: string) => void] {
    const { onChange } = props;

    const valueProps = DefineDefaultValue(props, "value", "defaultValue");
    const isControll = "value" in props;
    const [inputValue, setInputValue] = useState<string>(isShortTime(valueProps) ? valueProps : "");
    // 记录最后一次输入正确的时间字符串
    const lastRef = useRef(inputValue);

    function changeValue(val: string) {
        let _val = isShortTime(val) ? val : lastRef.current;
        if (val === "") {
            _val = "";
        }

        if (props.disabled) {
            return;
        }

        lastRef.current = _val;
        if (!isControll) {
            setInputValue(_val);
        }
        if (onChange) {
            onChange(_val);
        }
    }

    // 受控时候由外部更新输入框的值
    useEffect(() => {
        if (isControll) {
            setInputValue(props.value);
        }
    }, [isControll ? props.value : 1]);

    return [inputValue, setInputValue, changeValue];
}


src/interface.ts
import { AutoCompleteProps } from "xy-autocomplete/es/interface";

export interface TimeSelectProps extends AutoCompleteProps {
    /**
     * 时间
     */
    value?: string;
    /**
     * 默认时间
     */
    defaultValue?: string;
    /**
     * 根节点的附加类名
     */
    className?: string;
    /**
     * 内联样式
     */
    style?: React.CSSProperties;
    /**
     * 开始时间
     * @description 时间选择器开始时间范围, 比如 09:00
     */
    start?: string;
    /**
     * 结束时间
     * @description 时间选择器开始时间范围, 比如 18:00
     */
    end?: string;
    /**
     * 间隔时间
     * @description 默认 00:30
     */
    step?: string;
    /**
     * 最小可选时间
     * @description 最小时间之上的都看得见，但是禁用不可选的. 比如 06:00
     */
    min?: string;
    /**
     * 最大可选时间
     * @description 最小时间之下的都看得见，但是禁用不可选的 比如 17:00
     */
    max?: string;
    /**
     * hange事件
     */
    onChange?: (value: string) => void;
}


src/utils.ts
import { TimeSelectProps } from "./interface";
import { formatDate, isShortTime, shortTimeParse, shrotTimeTimestamp } from "utils-dom";
import { OptionConfig } from "xy-select/es/interface";

const LAST_SECOND = "23:59";
// (1 * 60 * 1000)为1秒的毫秒数
const SECOND_TIMESTAMP = 60 * 1000;

export function createTimeList(props: TimeSelectProps) {
    const { start, step = "00:30", min, max } = props;
    if ("dataSource" in props) {
        return props.dataSource;
    }
    const times: OptionConfig[] = [];
    const end = isShortTime(props.end) ? props.end : LAST_SECOND;
    const startTimestamp = shortTimeParse(isShortTime(start) ? start : "00:00").getTime();
    const endTimestamp = shortTimeParse(isShortTime(end) ? end : LAST_SECOND).getTime();
    const stepTimestamp = shrotTimeTimestamp(shortTimeParse(step));
    const minTimestamp = isShortTime(min) ? shortTimeParse(min).getTime() : 0;
    const maxTimestamp = isShortTime(max) ? shortTimeParse(max).getTime() : 0;

    // Tips: 修正最后 23:59 的时候需要等于 24:00 才进入此循环, (1 * 60 * 1000)为1秒的毫秒数，补齐24:00
    const isLastEnd = end === LAST_SECOND;
    const lastSecond = isLastEnd ? endTimestamp + SECOND_TIMESTAMP : endTimestamp;

    for (let time = startTimestamp; time <= lastSecond; time += stepTimestamp) {
        if (time === lastSecond && isLastEnd) {
            time -= SECOND_TIMESTAMP;
        }
        const date = new Date(time);
        times.push({
            label: formatDate(date, "HH:mm"),
            value: formatDate(date, "HH:mm"),
            disabled: (minTimestamp && time < minTimestamp) || (maxTimestamp && time > maxTimestamp)
        });
    }

    return times;
}
